src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/debug.cxx

```diff
@@ -254,64 +255,56 @@ namespace xclhwemhal2
 
-    unsigned int timeCheck = 0;
     unsigned int parseCount = 0;
-    while (inst && inst->get_simulator_started())
+    std::vector<std::string> lMatchedStrings = {"SIM-IPC's external process can be connected to instance",
+                                      "SystemC TLM functional mode",
+                                      "HLS_PRINT",
+                                      "Exiting xsim",
+                                      "FATAL_ERROR"};
+    xclemulation::sParseLog lParseLog(std::string(getSimPath() + "/simulate.log"), xclemulation::eEmulationType::eHw_Emu, lMatchedStrings);
+    while (get_simulator_started())
     {
       sleep(10);
-      if (inst->get_simulator_started() == false) {
-        return;
-      }
-
-      std::string consoleMsg = "INFO: [HW-EMU 07-0] Please refer the path \"" + inst->getSimPath() + "/simulate.log\" for more detailed simulation infos, errors and warnings.";
-      if (std::find(inst->parsedMsgs.begin(), inst->parsedMsgs.end(), consoleMsg) == inst->parsedMsgs.end()) {
-        inst->logMessage(consoleMsg);
-        inst->parsedMsgs.push_back(consoleMsg);
-      }
+      if (not get_simulator_started())
+       break;
 
       auto l_time_end = std::chrono::high_resolution_clock::now();
-      if (std::chrono::duration<double>(l_time_end - l_time).count() > 300) {
-        l_time = std::chrono::high_resolution_clock::now();
-        inst->mPrintMessagesLock.lock();
-        if (inst->get_simulator_started() == false)
-        {
-          inst->mPrintMessagesLock.unlock();
+      if (std::chrono::duration_cast<std::chrono::seconds>(l_time_end - l_time).count() > 300) {
+       
+        std::lock_guard<std::mutex> guard(mPrintMessagesLock);
+        if (get_simulator_started() == false)
           return;
-        }
-        inst->parseSimulateLog();
-        inst->fetchAndPrintMessages();
-        inst->mPrintMessagesLock.unlock();
+        
+        parseSimulateLog();
+        fetchAndPrintMessages();
       }
 
       auto end_time = std::chrono::high_resolution_clock::now();
-      if (std::chrono::duration<double>(end_time - start_time).count() > timeCheck) {
+      if (std::chrono::duration_cast<std::chrono::seconds>(end_time - start_time).count() <= 300) {
 
-        start_time = std::chrono::high_resolution_clock::now();
-        inst->mPrintMessagesLock.lock();
+        std::lock_guard<std::mutex> guard(mPrintMessagesLock);
 
-        if (inst->get_simulator_started() == false) {
-          inst->mPrintMessagesLock.unlock();
+        if (get_simulator_started() == false) 
           return;
-        }
 
-        inst->parseLog();
+        lParseLog.parseLog();
         parseCount++;
-
-        if (parseCount%5 == 0 && timeCheck < 300) {
-          timeCheck += 10;
+        if (parseCount%5 == 0) {
+            std::this_thread::sleep_for(std::chrono::seconds(10*(parseCount/5)));
         }
-
-        inst->mPrintMessagesLock.unlock();
       }
-    }
+      
+    } //while end.
   }
```

Comment:
I cannot comprehend this code.  There are no comments but bunch of parsing, comparison of random times, etc.
I am happy it is simplified in this PR, but still I have no clue.

It would help a lot to re-factor code that is deeply nested.  Outer while-loop should probably call separate re-factored function.  And no function should be longer than 7-10 lines in general.

Reply:
The function is updated with comments. Hoping it explains the behavior.